README.md
# Currencies

This library is a clone of the [JS version](https://github.com/paylike/currencies)

List of currencies supported by Paylike for transactions and accounts.

Read more:

- https://paylike.io/account/currency
- https://paylike.io/transactions/currencies

## Account currencies

These are the currencies you can use as a base for your account. Also known as
account, settlement and funding currency.

### Global

- EUR (Euro)
- USD (United States dollar)
- GBP (British pound sterling)

### Nordic

- DKK (Danish krone)
- NOK (Norwegian krone)
- SEK (Swedish krona)

### Other EU

- CHF (Swiss franc)
- HUF (Hungarian forint)
- PLN (Polish złoty)
- RON (Romanian leu)
- CZK (Czech koruna)
- BGN (Bulgarian lev)
- HRK (Croatian kuna)

## Usage

Within the Paylike ecosystem you should use the `CurrencyCode` enum to refer to a currency.

```dart
var currencies = PaylikeCurrencies();
var eur = currencies.byCode(CurrencyCode.EUR);

// lookup by code
currencies.byCode(CurrencyCode.AED);
	// { code: 'AED', currency: 'United Arab Emirates dirham', numeric: '784' }

// lookup CurrencyCode
currencies.getCurrencyCode('EUR');
	// CurrencyCode.EUR

// get List<PaylikeCurrency>
currencies.list();

// Convert between minor and major respecting the exponent
currencies.toMinor(CurrencyCode.DKK, 100.00);
	// 10000

currencies.toMajor(CurrencyCode.DKK, 10000);
	// 100.00
```


CHANGELOG.md
## 1.0.7

- Fixed issues with documentation

## 1.0.6

- Added support for deprecation
- Added deprecated currencies: MRO, STD, VES
- Added back VEF

## 1.0.5

- Fixed a minor pubspec issue

## 1.0.4

- Added new function to support getting currency code based on string code

## 1.0.3

- Improved code generation setup
- Removed VES from currencies

## 1.0.2

- Added support for funding
- Added support for listing

## 1.0.1

- Updated README

## 1.0.0

- Initial version.


test/paylike_currencies_test.dart
import 'package:paylike_currencies/paylike_currencies.dart';
import 'package:test/test.dart';

void main() {
  group('A group of tests', () {
    final currencies = PaylikeCurrencies();
    test('Should be able to find PaylikeCurrency based on code', () {
      var currency = currencies.byCode(CurrencyCode.EUR);
      expect(currency.code, 'EUR');
      expect(currency.currency, 'Euro');
      expect(currency.numeric, 978);
      expect(currency.exponent, 2);
      expect(currency.funding, true);
    });

    test('toMajor toMinor should work as expected', () {
      expect(currencies.toMajor(CurrencyCode.EUR, 100), 1);
      expect(currencies.toMinor(CurrencyCode.EUR, 1), 100);
      expect(currencies.toMajor(CurrencyCode.JPY, 1), 1);
      expect(currencies.toMinor(CurrencyCode.JPY, 1), 1);
      expect(currencies.toMajor(CurrencyCode.JPY, 0), 0);
      expect(currencies.toMinor(CurrencyCode.JPY, 0), 0);

      expect(currencies.toMinor(CurrencyCode.EUR, 19.9), 1990);
      expect(currencies.toMinor(CurrencyCode.EUR, 19.99), 1999);
      expect(currencies.toMinor(CurrencyCode.EUR, 19.01), 1901);

      try {
        currencies.byNumeric(2355);
        fail('should not be able to reach this');
      } catch (e) {
        expect(e is MissingCurrencyException, true);
        expect((e as MissingCurrencyException).numeric, 2355);
      }
      try {
        currencies.getCurrencyCode('EURO');
        fail('should not be able to reach this');
      } catch (e) {
        expect(e is MissingCurrencyException, true);
        expect((e as MissingCurrencyException).requestCode, 'EURO');
      }
    });
  });
}


tool/builder.dart
import 'dart:convert';
import 'dart:io';

import 'package:build/build.dart';
import 'package:paylike_currencies/src/paylike_currency.dart';
import 'package:source_gen/source_gen.dart';

// @nodoc
Builder currencyBuilder(BuilderOptions options) =>
    LibraryBuilder(CurrencyGenerator(), generatedExtension: '.currency.g.dart');

/// Generates currency collection
/// @nodoc
class CurrencyGenerator extends Generator {
  const CurrencyGenerator();
  @override
  Future<String> generate(LibraryReader library, BuildStep step) async {
    var f = File('./lib/src/currencies.json');
    var raw = f.readAsStringSync();
    List<dynamic> currencies = jsonDecode(raw);

    final outputEnum = <String>[
      '''/// Code used when referencing a currency
enum CurrencyCode {

        '''
    ];
    final outputClass = <String>[
      '''/// Collection of all available currencies by [CurrencyCode]
class PaylikeCurrencyCollection {
          static final Map<CurrencyCode, PaylikeCurrency> currencies = {
      
      '''
    ];
    for (var currency in currencies) {
      var parsed = PaylikeCurrency.fromJSON(currency);
      outputEnum.add('''
      ${parsed.code},
      ''');
      outputClass.add('''
        CurrencyCode.${parsed.code}: PaylikeCurrency('${parsed.code}', '${parsed.currency}', ${parsed.numeric}, ${parsed.exponent}, ${parsed.funding}, ${parsed.deprecated}),
      ''');
    }

    outputEnum.add('}');
    outputClass.add('};}');

    return '''
    import 'package:paylike_currencies/src/paylike_currency.dart';

    ${outputEnum.join('')}
    ${outputClass.join('')}
    ''';
  }
}


lib/src/paylike_currency.dart
/// PaylikeCurrency describes a single currency in the collection
class PaylikeCurrency {
  /// Currency code (e.g. USD, EUR, GBP)
  String code;

  /// Currency name (e.g. United States dollar, Euro, Pound Sterling)
  String currency;

  /// Numeric code for the currency (e.g. 840, 978, 826)
  int numeric;

  /// Exponent to use with a given currency
  int exponent;

  /// Indicates currency support for your merchant account
  bool funding;

  /// Indicates if the given currency is supported yet (e.g. the old Venezuelan bolívar soberan is deprecated now)
  bool deprecated;

  /// Creates a PaylikeCurrency instance
  PaylikeCurrency(this.code, this.currency, this.numeric, this.exponent,
      this.funding, this.deprecated);

  /// Creates a PaylikeCurrency instance using JSON
  PaylikeCurrency.fromJSON(Map<String, dynamic> json)
      : code = json['code'],
        currency = json['currency'],
        numeric = int.parse(json['numeric']),
        exponent = json['exponent'],
        funding = json['funding'] != null,
        deprecated = json['deprecated'] != null;
}


lib/src/paylike_currencies_base.dart
import 'package:paylike_currencies/paylike_currencies.dart';
import 'dart:math';

/// Describes that a given currency the user is looking for
/// is missing from the collection
class MissingCurrencyException implements Exception {
  final String cause = 'Currency is missing';
  CurrencyCode? code;
  int? numeric;
  String? requestCode;
  MissingCurrencyException.fromNumeric(this.numeric);
  MissingCurrencyException.fromCode(this.code);
  MissingCurrencyException.fromStringCode(this.requestCode);
}

/// PaylikeCurrencies is responsible for currency related operations
class PaylikeCurrencies {
  /// Returns the CurrencyCode based on [code]
  CurrencyCode getCurrencyCode(String code) {
    return PaylikeCurrencyCollection.currencies.entries
        .firstWhere((element) => element.value.code == code,
            orElse: () => throw MissingCurrencyException.fromStringCode(code))
        .key;
  }

  /// Provides a [PaylikeCurrency] based on [code]
  PaylikeCurrency byCode(CurrencyCode code) {
    var element = PaylikeCurrencyCollection.currencies[code];
    if (element == null) {
      throw MissingCurrencyException.fromCode(code);
    }
    return element;
  }

  /// Lists all available currencies
  List<PaylikeCurrency> list() => List.from(
      PaylikeCurrencyCollection.currencies.entries.map((e) => e.value));

  /// Provides a [PaylikeCurrency] based on [numeric]
  PaylikeCurrency byNumeric(int numeric) {
    var element = PaylikeCurrencyCollection.currencies.entries.firstWhere(
        (element) => element.value.numeric == numeric,
        orElse: () => throw MissingCurrencyException.fromNumeric(numeric));
    return element.value;
  }

  /// Converts a currency to minor value using its exponent
  /// e.g: 1 USD to cents
  num toMinor(CurrencyCode code, num major) =>
      (major * pow(10, byCode(code).exponent)).round();

  /// Converts a currency to major value using its exponent
  /// e.g: 100 cents to 1 USD
  num toMajor(CurrencyCode code, num minor) =>
      minor / pow(10, byCode(code).exponent);
}


lib/src/paylike_currencies_base.currency.g.dart
// GENERATED CODE - DO NOT MODIFY BY HAND

// **************************************************************************
// CurrencyGenerator
// **************************************************************************

import 'package:paylike_currencies/src/paylike_currency.dart';

/// Code used when referencing a currency
enum CurrencyCode {
  AED,
  AFN,
  ALL,
  AMD,
  ANG,
  AOA,
  ARS,
  AUD,
  AWG,
  AZN,
  BAM,
  BBD,
  BDT,
  BGN,
  BHD,
  BIF,
  BMD,
  BND,
  BOB,
  BRL,
  BSD,
  BTN,
  BWP,
  BYR,
  BZD,
  CAD,
  CDF,
  CHF,
  CLP,
  CNY,
  COP,
  CRC,
  CUP,
  CVE,
  CZK,
  DJF,
  DKK,
  DOP,
  DZD,
  EGP,
  ERN,
  ETB,
  EUR,
  FJD,
  FKP,
  GBP,
  GEL,
  GHS,
  GIP,
  GMD,
  GNF,
  GTQ,
  GYD,
  HKD,
  HNL,
  HRK,
  HTG,
  HUF,
  IDR,
  ILS,
  INR,
  IQD,
  IRR,
  ISK,
  JMD,
  JOD,
  JPY,
  KES,
  KGS,
  KHR,
  KMF,
  KPW,
  KRW,
  KWD,
  KYD,
  KZT,
  LAK,
  LBP,
  LKR,
  LRD,
  LSL,
  MAD,
  MDL,
  MGA,
  MKD,
  MMK,
  MNT,
  MOP,
  MRU,
  MRO,
  MUR,
  MVR,
  MWK,
  MXN,
  MYR,
  MZN,
  NAD,
  NGN,
  NIO,
  NOK,
  NPR,
  NZD,
  OMR,
  PAB,
  PEN,
  PGK,
  PHP,
  PKR,
  PLN,
  PYG,
  QAR,
  RON,
  RSD,
  RUB,
  RWF,
  SAR,
  SBD,
  SCR,
  SDG,
  SEK,
  SGD,
  SHP,
  SLL,
  SOS,
  SRD,
  STN,
  STD,
  SYP,
  SZL,
  THB,
  TJS,
  TMT,
  TND,
  TOP,
  TRY,
  TTD,
  TWD,
  TZS,
  UAH,
  UGX,
  USD,
  UYU,
  UZS,
  VEF,
  VES,
  VND,
  VUV,
  WST,
  XAF,
  XCD,
  XOF,
  XPF,
  YER,
  ZAR,
  ZMK,
  ZWL,
}

/// Collection of all available currencies by [CurrencyCode]
class PaylikeCurrencyCollection {
  static final Map<CurrencyCode, PaylikeCurrency> currencies = {
    CurrencyCode.AED: PaylikeCurrency(
        'AED', 'United Arab Emirates dirham', 784, 2, false, false),
    CurrencyCode.AFN:
        PaylikeCurrency('AFN', 'Afghan afghani', 971, 2, false, false),
    CurrencyCode.ALL:
        PaylikeCurrency('ALL', 'Albanian lek', 8, 2, false, false),
    CurrencyCode.AMD:
        PaylikeCurrency('AMD', 'Armenian dram', 51, 2, false, false),
    CurrencyCode.ANG: PaylikeCurrency(
        'ANG', 'Netherlands Antillean guilder', 532, 2, false, false),
    CurrencyCode.AOA:
        PaylikeCurrency('AOA', 'Angolan kwanza', 973, 2, false, false),
    CurrencyCode.ARS:
        PaylikeCurrency('ARS', 'Argentine peso', 32, 2, false, false),
    CurrencyCode.AUD:
        PaylikeCurrency('AUD', 'Australian dollar', 36, 2, false, false),
    CurrencyCode.AWG:
        PaylikeCurrency('AWG', 'Aruban florin', 533, 2, false, false),
    CurrencyCode.AZN:
        PaylikeCurrency('AZN', 'Azerbaijani manat', 944, 2, false, false),
    CurrencyCode.BAM: PaylikeCurrency(
        'BAM', 'Bosnia and Herzegovina convertible mark', 977, 2, false, false),
    CurrencyCode.BBD:
        PaylikeCurrency('BBD', 'Barbados dollar', 52, 2, false, false),
    CurrencyCode.BDT:
        PaylikeCurrency('BDT', 'Bangladeshi taka', 50, 2, false, false),
    CurrencyCode.BGN:
        PaylikeCurrency('BGN', 'Bulgarian lev', 975, 2, true, false),
    CurrencyCode.BHD:
        PaylikeCurrency('BHD', 'Bahraini dinar', 48, 3, false, false),
    CurrencyCode.BIF:
        PaylikeCurrency('BIF', 'Burundian franc', 108, 0, false, false),
    CurrencyCode.BMD:
        PaylikeCurrency('BMD', 'Bermudian dollar', 60, 2, false, false),
    CurrencyCode.BND:
        PaylikeCurrency('BND', 'Brunei dollar', 96, 2, false, false),
    CurrencyCode.BOB: PaylikeCurrency('BOB', 'Boliviano', 68, 2, false, false),
    CurrencyCode.BRL:
        PaylikeCurrency('BRL', 'Brazilian real', 986, 2, false, false),
    CurrencyCode.BSD:
        PaylikeCurrency('BSD', 'Bahamian dollar', 44, 2, false, false),
    CurrencyCode.BTN:
        PaylikeCurrency('BTN', 'Bhutanese ngultrum', 64, 2, false, false),
    CurrencyCode.BWP:
        PaylikeCurrency('BWP', 'Botswana pula', 72, 2, false, false),
    CurrencyCode.BYR:
        PaylikeCurrency('BYR', 'Belarusian ruble', 974, 0, false, false),
    CurrencyCode.BZD:
        PaylikeCurrency('BZD', 'Belize dollar', 84, 2, false, false),
    CurrencyCode.CAD:
        PaylikeCurrency('CAD', 'Canadian dollar', 124, 2, false, false),
    CurrencyCode.CDF:
        PaylikeCurrency('CDF', 'Congolese franc', 976, 2, false, false),
    CurrencyCode.CHF:
        PaylikeCurrency('CHF', 'Swiss franc', 756, 2, true, false),
    CurrencyCode.CLP:
        PaylikeCurrency('CLP', 'Chilean peso', 152, 0, false, false),
    CurrencyCode.CNY:
        PaylikeCurrency('CNY', 'Chinese yuan', 156, 2, false, false),
    CurrencyCode.COP:
        PaylikeCurrency('COP', 'Colombian peso', 170, 2, false, false),
    CurrencyCode.CRC:
        PaylikeCurrency('CRC', 'Costa Rican colon', 188, 2, false, false),
    CurrencyCode.CUP:
        PaylikeCurrency('CUP', 'Cuban peso', 192, 2, false, false),
    CurrencyCode.CVE:
        PaylikeCurrency('CVE', 'Cape Verde escudo', 132, 2, false, false),
    CurrencyCode.CZK:
        PaylikeCurrency('CZK', 'Czech koruna', 203, 2, true, false),
    CurrencyCode.DJF:
        PaylikeCurrency('DJF', 'Djiboutian franc', 262, 0, false, false),
    CurrencyCode.DKK:
        PaylikeCurrency('DKK', 'Danish krone', 208, 2, true, false),
    CurrencyCode.DOP:
        PaylikeCurrency('DOP', 'Dominican peso', 214, 2, false, false),
    CurrencyCode.DZD:
        PaylikeCurrency('DZD', 'Algerian dinar', 12, 2, false, false),
    CurrencyCode.EGP:
        PaylikeCurrency('EGP', 'Egyptian pound', 818, 2, false, false),
    CurrencyCode.ERN:
        PaylikeCurrency('ERN', 'Eritrean nakfa', 232, 2, false, false),
    CurrencyCode.ETB:
        PaylikeCurrency('ETB', 'Ethiopian birr', 230, 2, false, false),
    CurrencyCode.EUR: PaylikeCurrency('EUR', 'Euro', 978, 2, true, false),
    CurrencyCode.FJD:
        PaylikeCurrency('FJD', 'Fiji dollar', 242, 2, false, false),
    CurrencyCode.FKP:
        PaylikeCurrency('FKP', 'Falkland Islands pound', 238, 2, false, false),
    CurrencyCode.GBP:
        PaylikeCurrency('GBP', 'Pound sterling', 826, 2, true, false),
    CurrencyCode.GEL:
        PaylikeCurrency('GEL', 'Georgian lari', 981, 2, false, false),
    CurrencyCode.GHS:
        PaylikeCurrency('GHS', 'Ghanaian cedi', 936, 2, false, false),
    CurrencyCode.GIP:
        PaylikeCurrency('GIP', 'Gibraltar pound', 292, 2, false, false),
    CurrencyCode.GMD:
        PaylikeCurrency('GMD', 'Gambian dalasi', 270, 2, false, false),
    CurrencyCode.GNF:
        PaylikeCurrency('GNF', 'Guinean franc', 324, 0, false, false),
    CurrencyCode.GTQ:
        PaylikeCurrency('GTQ', 'Guatemalan quetzal', 320, 2, false, false),
    CurrencyCode.GYD:
        PaylikeCurrency('GYD', 'Guyanese dollar', 328, 2, false, false),
    CurrencyCode.HKD:
        PaylikeCurrency('HKD', 'Hong Kong dollar', 344, 2, false, false),
    CurrencyCode.HNL:
        PaylikeCurrency('HNL', 'Honduran lempira', 340, 2, false, false),
    CurrencyCode.HRK:
        PaylikeCurrency('HRK', 'Croatian kuna', 191, 2, true, false),
    CurrencyCode.HTG:
        PaylikeCurrency('HTG', 'Haitian gourde', 332, 2, false, false),
    CurrencyCode.HUF:
        PaylikeCurrency('HUF', 'Hungarian forint', 348, 2, true, false),
    CurrencyCode.IDR:
        PaylikeCurrency('IDR', 'Indonesian rupiah', 360, 2, false, false),
    CurrencyCode.ILS:
        PaylikeCurrency('ILS', 'Israeli new shekel', 376, 2, false, false),
    CurrencyCode.INR:
        PaylikeCurrency('INR', 'Indian rupee', 356, 2, false, false),
    CurrencyCode.IQD:
        PaylikeCurrency('IQD', 'Iraqi dinar', 368, 3, false, false),
    CurrencyCode.IRR:
        PaylikeCurrency('IRR', 'Iranian rial', 364, 2, false, false),
    CurrencyCode.ISK:
        PaylikeCurrency('ISK', 'Icelandic króna', 352, 2, false, false),
    CurrencyCode.JMD:
        PaylikeCurrency('JMD', 'Jamaican dollar', 388, 2, false, false),
    CurrencyCode.JOD:
        PaylikeCurrency('JOD', 'Jordanian dinar', 400, 3, false, false),
    CurrencyCode.JPY:
        PaylikeCurrency('JPY', 'Japanese yen', 392, 0, false, false),
    CurrencyCode.KES:
        PaylikeCurrency('KES', 'Kenyan shilling', 404, 2, false, false),
    CurrencyCode.KGS:
        PaylikeCurrency('KGS', 'Kyrgyzstani som', 417, 2, false, false),
    CurrencyCode.KHR:
        PaylikeCurrency('KHR', 'Cambodian riel', 116, 2, false, false),
    CurrencyCode.KMF:
        PaylikeCurrency('KMF', 'Comoro franc', 174, 0, false, false),
    CurrencyCode.KPW:
        PaylikeCurrency('KPW', 'North Korean won', 408, 2, false, false),
    CurrencyCode.KRW:
        PaylikeCurrency('KRW', 'South Korean won', 410, 0, false, false),
    CurrencyCode.KWD:
        PaylikeCurrency('KWD', 'Kuwaiti dinar', 414, 3, false, false),
    CurrencyCode.KYD:
        PaylikeCurrency('KYD', 'Cayman Islands dollar', 136, 2, false, false),
    CurrencyCode.KZT:
        PaylikeCurrency('KZT', 'Kazakhstani tenge', 398, 2, false, false),
    CurrencyCode.LAK: PaylikeCurrency('LAK', 'Lao kip', 418, 2, false, false),
    CurrencyCode.LBP:
        PaylikeCurrency('LBP', 'Lebanese pound', 422, 2, false, false),
    CurrencyCode.LKR:
        PaylikeCurrency('LKR', 'Sri Lankan rupee', 144, 2, false, false),
    CurrencyCode.LRD:
        PaylikeCurrency('LRD', 'Liberian dollar', 430, 2, false, false),
    CurrencyCode.LSL:
        PaylikeCurrency('LSL', 'Lesotho loti', 426, 2, false, false),
    CurrencyCode.MAD:
        PaylikeCurrency('MAD', 'Moroccan dirham', 504, 2, false, false),
    CurrencyCode.MDL:
        PaylikeCurrency('MDL', 'Moldovan leu', 498, 2, false, false),
    CurrencyCode.MGA:
        PaylikeCurrency('MGA', 'Malagasy ariary', 969, 2, false, false),
    CurrencyCode.MKD:
        PaylikeCurrency('MKD', 'Macedonian denar', 807, 2, false, false),
    CurrencyCode.MMK:
        PaylikeCurrency('MMK', 'Myanmar kyat', 104, 2, false, false),
    CurrencyCode.MNT:
        PaylikeCurrency('MNT', 'Mongolian tögrög', 496, 2, false, false),
    CurrencyCode.MOP:
        PaylikeCurrency('MOP', 'Macanese pataca', 446, 2, false, false),
    CurrencyCode.MRU:
        PaylikeCurrency('MRU', 'Mauritanian ouguiya', 929, 2, false, false),
    CurrencyCode.MRO:
        PaylikeCurrency('MRO', 'Mauritanian ouguiya', 478, 2, false, true),
    CurrencyCode.MUR:
        PaylikeCurrency('MUR', 'Mauritian rupee', 480, 2, false, false),
    CurrencyCode.MVR:
        PaylikeCurrency('MVR', 'Maldivian rufiyaa', 462, 2, false, false),
    CurrencyCode.MWK:
        PaylikeCurrency('MWK', 'Malawian kwacha', 454, 2, false, false),
    CurrencyCode.MXN:
        PaylikeCurrency('MXN', 'Mexican peso', 484, 2, false, false),
    CurrencyCode.MYR:
        PaylikeCurrency('MYR', 'Malaysian ringgit', 458, 2, false, false),
    CurrencyCode.MZN:
        PaylikeCurrency('MZN', 'Mozambican metical', 943, 2, false, false),
    CurrencyCode.NAD:
        PaylikeCurrency('NAD', 'Namibian dollar', 516, 2, false, false),
    CurrencyCode.NGN:
        PaylikeCurrency('NGN', 'Nigerian naira', 566, 2, false, false),
    CurrencyCode.NIO:
        PaylikeCurrency('NIO', 'Nicaraguan córdoba', 558, 2, false, false),
    CurrencyCode.NOK:
        PaylikeCurrency('NOK', 'Norwegian krone', 578, 2, true, false),
    CurrencyCode.NPR:
        PaylikeCurrency('NPR', 'Nepalese rupee', 524, 2, false, false),
    CurrencyCode.NZD:
        PaylikeCurrency('NZD', 'New Zealand dollar', 554, 2, false, false),
    CurrencyCode.OMR:
        PaylikeCurrency('OMR', 'Omani rial', 512, 3, false, false),
    CurrencyCode.PAB:
        PaylikeCurrency('PAB', 'Panamanian balboa', 590, 2, false, false),
    CurrencyCode.PEN:
        PaylikeCurrency('PEN', 'Peruvian Sol', 604, 2, false, false),
    CurrencyCode.PGK:
        PaylikeCurrency('PGK', 'Papua New Guinean kina', 598, 2, false, false),
    CurrencyCode.PHP:
        PaylikeCurrency('PHP', 'Philippine peso', 608, 2, false, false),
    CurrencyCode.PKR:
        PaylikeCurrency('PKR', 'Pakistani rupee', 586, 2, false, false),
    CurrencyCode.PLN:
        PaylikeCurrency('PLN', 'Polish złoty', 985, 2, true, false),
    CurrencyCode.PYG:
        PaylikeCurrency('PYG', 'Paraguayan guaraní', 600, 0, false, false),
    CurrencyCode.QAR:
        PaylikeCurrency('QAR', 'Qatari riyal', 634, 2, false, false),
    CurrencyCode.RON:
        PaylikeCurrency('RON', 'Romanian leu', 946, 2, true, false),
    CurrencyCode.RSD:
        PaylikeCurrency('RSD', 'Serbian dinar', 941, 2, false, false),
    CurrencyCode.RUB:
        PaylikeCurrency('RUB', 'Russian ruble', 643, 2, false, false),
    CurrencyCode.RWF:
        PaylikeCurrency('RWF', 'Rwandan franc', 646, 0, false, false),
    CurrencyCode.SAR:
        PaylikeCurrency('SAR', 'Saudi riyal', 682, 2, false, false),
    CurrencyCode.SBD:
        PaylikeCurrency('SBD', 'Solomon Islands dollar', 90, 2, false, false),
    CurrencyCode.SCR:
        PaylikeCurrency('SCR', 'Seychelles rupee', 690, 2, false, false),
    CurrencyCode.SDG:
        PaylikeCurrency('SDG', 'Sudanese pound', 938, 2, false, false),
    CurrencyCode.SEK:
        PaylikeCurrency('SEK', 'Swedish krona', 752, 2, true, false),
    CurrencyCode.SGD:
        PaylikeCurrency('SGD', 'Singapore dollar', 702, 2, false, false),
    CurrencyCode.SHP:
        PaylikeCurrency('SHP', 'Saint Helena pound', 654, 2, false, false),
    CurrencyCode.SLL:
        PaylikeCurrency('SLL', 'Sierra Leonean leone', 694, 2, false, false),
    CurrencyCode.SOS:
        PaylikeCurrency('SOS', 'Somali shilling', 706, 2, false, false),
    CurrencyCode.SRD:
        PaylikeCurrency('SRD', 'Surinamese dollar', 968, 2, false, false),
    CurrencyCode.STN: PaylikeCurrency(
        'STN', 'São Tomé and Príncipe dobra', 930, 2, false, false),
    CurrencyCode.STD: PaylikeCurrency(
        'STD', 'São Tomé and Príncipe dobra', 678, 2, false, true),
    CurrencyCode.SYP:
        PaylikeCurrency('SYP', 'Syrian pound', 760, 2, false, false),
    CurrencyCode.SZL:
        PaylikeCurrency('SZL', 'Swazi lilangeni', 748, 2, false, false),
    CurrencyCode.THB: PaylikeCurrency('THB', 'Thai baht', 764, 2, false, false),
    CurrencyCode.TJS:
        PaylikeCurrency('TJS', 'Tajikistani somoni', 972, 2, false, false),
    CurrencyCode.TMT:
        PaylikeCurrency('TMT', 'Turkmenistani manat', 934, 2, false, false),
    CurrencyCode.TND:
        PaylikeCurrency('TND', 'Tunisian dinar', 788, 3, false, false),
    CurrencyCode.TOP:
        PaylikeCurrency('TOP', 'Tongan paʻanga', 776, 2, false, false),
    CurrencyCode.TRY:
        PaylikeCurrency('TRY', 'Turkish lira', 949, 2, false, false),
    CurrencyCode.TTD: PaylikeCurrency(
        'TTD', 'Trinidad and Tobago dollar', 780, 2, false, false),
    CurrencyCode.TWD:
        PaylikeCurrency('TWD', 'New Taiwan dollar', 901, 2, false, false),
    CurrencyCode.TZS:
        PaylikeCurrency('TZS', 'Tanzanian shilling', 834, 2, false, false),
    CurrencyCode.UAH:
        PaylikeCurrency('UAH', 'Ukrainian hryvnia', 980, 2, false, false),
    CurrencyCode.UGX:
        PaylikeCurrency('UGX', 'Ugandan shilling', 800, 0, false, false),
    CurrencyCode.USD:
        PaylikeCurrency('USD', 'United States dollar', 840, 2, true, false),
    CurrencyCode.UYU:
        PaylikeCurrency('UYU', 'Uruguayan peso', 858, 2, false, false),
    CurrencyCode.UZS:
        PaylikeCurrency('UZS', 'Uzbekistan som', 860, 2, false, false),
    CurrencyCode.VEF:
        PaylikeCurrency('VEF', 'Venezuelan bolívar', 937, 2, false, false),
    CurrencyCode.VES: PaylikeCurrency(
        'VES', 'Venezuelan bolívar soberano', 928, 2, false, true),
    CurrencyCode.VND:
        PaylikeCurrency('VND', 'Vietnamese dong', 704, 0, false, false),
    CurrencyCode.VUV:
        PaylikeCurrency('VUV', 'Vanuatu vatu', 548, 0, false, false),
    CurrencyCode.WST:
        PaylikeCurrency('WST', 'Samoan tala', 882, 2, false, false),
    CurrencyCode.XAF:
        PaylikeCurrency('XAF', 'CFA franc BEAC', 950, 0, false, false),
    CurrencyCode.XCD:
        PaylikeCurrency('XCD', 'East Caribbean dollar', 951, 2, false, false),
    CurrencyCode.XOF:
        PaylikeCurrency('XOF', 'CFA franc BCEAO', 952, 0, false, false),
    CurrencyCode.XPF: PaylikeCurrency('XPF', 'CFP franc', 953, 0, false, false),
    CurrencyCode.YER:
        PaylikeCurrency('YER', 'Yemeni rial', 886, 2, false, false),
    CurrencyCode.ZAR:
        PaylikeCurrency('ZAR', 'South African rand', 710, 2, false, false),
    CurrencyCode.ZMK:
        PaylikeCurrency('ZMK', 'Zambian kwacha', 894, 2, false, false),
    CurrencyCode.ZWL:
        PaylikeCurrency('ZWL', 'Zimbabwean dollar', 716, 2, false, false),
  };
}
